### PR TITLE
Fix unaligned memory load, store operations

### PR DESCRIPTION
**Pull request by gergocs**
Signed-off-by: Gergo Csizi gergocs@inf.u-szeged.hu

### CODE REVIEW

File: src/jit/MemoryInl.h [hunk 1]
```diff
@@ -378,8 +378,18 @@ static void emitLoad(sljit_compiler* compiler, Instruction* instr)
         JITArg valueArg;
         floatOperandToArg(compiler, operands + 1, valueArg, SLJIT_FR0);
 
-        // TODO: sljit_emit_fmem for unaligned access
-        sljit_emit_fop1(compiler, opcode, valueArg.arg, valueArg.argw, addr.memArg.arg, addr.memArg.argw);
+
+        if (addr.memArg.argw != 0) {
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:382]: I think argw is not enough to check this, you would need to compute `base + offset`. It is only enough for immediate offsets

File: src/jit/MemoryInl.h [hunk 2]
```diff
@@ -701,8 +711,12 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
             addr.loadArg.argw = 0;
         }
 
-        // TODO: sljit_emit_fmem for unaligned access
-        sljit_emit_fop1(compiler, opcode, addr.memArg.arg, addr.memArg.argw, addr.loadArg.arg, addr.loadArg.argw);
+        if (addr.memArg.argw != 0) {
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:714]: Same problem here.

File: src/jit/MemoryInl.h [hunk 3]
```diff
@@ -378,8 +378,18 @@ static void emitLoad(sljit_compiler* compiler, Instruction* instr)
         JITArg valueArg;
         floatOperandToArg(compiler, operands + 1, valueArg, SLJIT_FR0);
 
-        // TODO: sljit_emit_fmem for unaligned access
-        sljit_emit_fop1(compiler, opcode, valueArg.arg, valueArg.argw, addr.memArg.arg, addr.memArg.argw);
+
+        if (addr.memArg.argw != 0) {
+            sljit_s32 tmpReg = GET_TARGET_REG(valueArg.arg, SLJIT_FR1);
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:383]: Why FR1 and not FR0?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/MemoryInl.h (hunk 1 above)
Comment (gergocs) [reply src/jit/MemoryInl.h:382]: The goal was only the immediate case.

File: src/jit/MemoryInl.h (hunk 3 above)
Comment (zherczeg) [reply src/jit/MemoryInl.h:386]: Used for what? This is a load operation

File: src/jit/MemoryInl.h [hunk 4]
```diff
@@ -378,8 +378,21 @@ static void emitLoad(sljit_compiler* compiler, Instruction* instr)
         JITArg valueArg;
         floatOperandToArg(compiler, operands + 1, valueArg, SLJIT_FR0);
 
-        // TODO: sljit_emit_fmem for unaligned access
+
+#if (defined SLJIT_FPU_UNALIGNED && SLJIT_FPU_UNALIGNED)
         sljit_emit_fop1(compiler, opcode, valueArg.arg, valueArg.argw, addr.memArg.arg, addr.memArg.argw);
+#else
+        if (!SLJIT_IS_IMM(addr.memArg.arg)) {
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:385]: How can a memarg be immediate?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/MemoryInl.h [hunk 5]
```diff
@@ -378,8 +378,17 @@ static void emitLoad(sljit_compiler* compiler, Instruction* instr)
...
-        // TODO: sljit_emit_fmem for unaligned access
+
+#if (defined SLJIT_FPU_UNALIGNED && SLJIT_FPU_UNALIGNED)
         sljit_emit_fop1(compiler, opcode, valueArg.arg, valueArg.argw, addr.memArg.arg, addr.memArg.argw);
+#else /* SLJIT_FPU_UNALIGNED */
+        sljit_s32 tmpReg = GET_TARGET_REG(valueArg.arg, SLJIT_FR0);
+        sljit_emit_fmem(compiler, opcode | SLJIT_MEM_UNALIGNED, tmpReg, addr.memArg.arg, addr.memArg.argw);
+
+        if (SLJIT_IS_MEM(valueArg.arg)) {
+            sljit_emit_fmem(compiler, opcode | SLJIT_MEM_STORE | SLJIT_MEM_UNALIGNED, tmpReg, valueArg.arg, valueArg.argw);
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:389]: This should be sljit_emit_fop1 since the value is aligned, isn't it?

File: src/jit/MemoryInl.h [hunk 6]
```diff
@@ -701,8 +710,12 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
             addr.loadArg.argw = 0;
         }
 
-        // TODO: sljit_emit_fmem for unaligned access
+#if (defined SLJIT_FPU_UNALIGNED && SLJIT_FPU_UNALIGNED)
         sljit_emit_fop1(compiler, opcode, addr.memArg.arg, addr.memArg.argw, addr.loadArg.arg, addr.loadArg.argw);
+#else /* SLJIT_FPU_UNALIGNED */
+        sljit_emit_fop1(compiler, opcode | SLJIT_MEM_UNALIGNED, addr.memArg.arg, addr.memArg.argw, addr.loadArg.arg, addr.loadArg.argw);
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:716]: sljit_emit_fop1 does not support SLJIT_MEM_UNALIGNED

File: src/jit/MemoryInl.h [hunk 7]
```diff
@@ -739,8 +752,7 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
             sljit_emit_op1(compiler, SLJIT_MOV, dstReg1, 0, SLJIT_IMM, valueArgPair.arg1w);
 #endif /* SLJIT_BIG_ENDIAN */
         }
-
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:742]: Do not remove this newline

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Comment by zherczeg** [comment]
Could you update this patch?

**Comment by zherczeg** [comment]
Please update this patch.

File: src/jit/MemoryInl.h [hunk 8]
```diff
@@ -378,8 +378,17 @@ static void emitLoad(sljit_compiler* compiler, Instruction* instr)
         JITArg valueArg;
         floatOperandToArg(compiler, operands + 1, valueArg, SLJIT_FR0);
 
-        // TODO: sljit_emit_fmem for unaligned access
+
```
Comment (zherczeg) [inline src/jit/MemoryInl.h:381]: This newline is not needed.

**Review by zherczeg — CHANGES_REQUESTED** [review]
Good patch. Only a small change is needed.

**Review by zherczeg — APPROVED** [review]
LGTM

**Pull request merged**